apps/src/lab2/views/components/Instructions/NavigationArea.tsx

```diff
@@ -112,28 +115,99 @@ const NavigationArea: React.FC<NavigationAreaProps> = ({
 
-  const canShowButton = useMemo(() => {
+  // Certain labs hide the continue button until the user has met the conditions for the next level. Example: Music lab.
+  const isLabHidesContinueButton = useMemo(() => {
+    return LABS_HIDE_CONTINUE_BUTTON_RESOURCE_PANEL.includes(appName);
+  }, [appName]);
+
+  // For labs that hide the continue button, we show the button if the user has met the conditions for the next level.
+  // For labs that always show the continue button, this boolean is always true.
+  const showContinueButton = useMemo(() => {
+    let showContinueButton = true;
+    if (isLabHidesContinueButton) {
+      if (isPredictLevel) {
+        showContinueButton = predictResponseSubmitted;
+      } else if (hasValidationConditions) {
+        showContinueButton = validationSatisfied;
+      } else {
+        showContinueButton = hasRun; // We are assuming that the lab does not use requireRun as in the case of music lab.
+      }
+    }
+    return showContinueButton;
+  }, [
+    isLabHidesContinueButton,
+    isPredictLevel,
+    predictResponseSubmitted,
+    hasValidationConditions,
+    validationSatisfied,
+    hasRun,
+  ]);
+
+  // For labs that always show the continue button, the button is enabled if the user has met the conditions for the next level.
+  // For labs that hide the continue button, this boolean is always true.
+  const continueButtonIsEnabled = useMemo(() => {
+    if (isLabHidesContinueButton) {
+      return true;
+    }
     if (isPredictLevel) {
       return predictResponseSubmitted;
-    } else if (submittable && hasSubmitted) {
-      return true;
     } else if (hasValidationConditions) {
       return validationSatisfied;
+    } else if (requireRun) {
+      return hasRun;
     } else {
-      return submittable || !requireRun || hasRun;
+      return true;
     }
   }, [
     isPredictLevel,
-    hasValidationConditions,
     predictResponseSubmitted,
+    hasValidationConditions,
     validationSatisfied,
+    requireRun,
+    hasRun,
+    isLabHidesContinueButton,
+  ]);
+
+  const continueTooltipMessage = useMemo(() => {
+    if (submittable) {
+      return undefined;
+    }
+    if (isPredictLevel) {
+      return hasNextLevel
+        ? lab2I18n.toContinueSubmitPrediction()
+        : lab2I18n.toFinishSubmitPrediction();
+    } else if (hasValidationConditions && !validationSatisfied) {
+      return hasNextLevel
+        ? lab2I18n.toContinueValidate()
+        : lab2I18n.toFinishValidate();
+    } else if (requireRun && !hasRun) {
+      return hasNextLevel ? lab2I18n.toContinueRun() : lab2I18n.toFinishRun();
+    }
+    return undefined;
+  }, [
     submittable,
-    hasSubmitted,
+    hasNextLevel,
+    isPredictLevel,
+    hasValidationConditions,
+    validationSatisfied,
     requireRun,
     hasRun,
   ]);
 
-  if (!submittable && !canShowButton && !feedbackMessage) {
+  const submitButtonEnabled =
+    disableEditRunForSubmission || hasSubmitted || (hasRun && hasEdited);
```

Comment:
Can we just pull this logic into `SubmitButton` instead of passing through the tooltip prop?

Reply:
Done!